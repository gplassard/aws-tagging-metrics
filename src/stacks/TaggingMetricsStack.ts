import path from 'path';
import { Duration, RemovalPolicy, Stack, StackProps } from 'aws-cdk-lib';
import { Effect, PolicyStatement } from 'aws-cdk-lib/aws-iam';
import { Architecture } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { LogGroup, RetentionDays } from 'aws-cdk-lib/aws-logs';
import { Construct } from 'constructs';

export interface Props extends StackProps {
  baseName: string;
  tags: {
    Env: string;
    Team: string;
  };
}

export class TaggingMetricsStack extends Stack {
  constructor(scope: Construct, id: string, props: Props) {
    super(scope, id, {
      stackName: `${props.baseName}-${props.tags.Env}`,
      tags: props.tags,
    });

    const functionName = `${props.baseName}-summary-metrics-exporter-${props.tags.Env}`;
    const summaryMetricsExporter = new NodejsFunction(this, 'summary-metrics-exporter', {
      functionName,
      entry: path.join(__dirname, '../lambdas/summary-metrics/handler.ts'),
      architecture: Architecture.ARM_64,
      timeout: Duration.seconds(60),
      memorySize: 512,
      initialPolicy: [
        new PolicyStatement({
          effect: Effect.ALLOW,
          actions: ['tag:GetResources'],
          resources: ['*'],
        }),
      ],
    });
    const logGroup = new LogGroup(this, 'summary-metrics-exporter-logs', {
      logGroupName: `/aws/lambda/${functionName}`,
      retention: RetentionDays.ONE_WEEK,
      removalPolicy: RemovalPolicy.DESTROY,
    });
    summaryMetricsExporter.node.addDependency(logGroup);
  }
}
