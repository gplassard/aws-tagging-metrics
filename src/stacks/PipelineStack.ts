import { App, CfnDynamicReference, CfnDynamicReferenceService, SecretValue, Stack, StackProps, Stage } from 'aws-cdk-lib';
import { CodePipeline, CodePipelineSource, ShellStep } from 'aws-cdk-lib/pipelines';
import { TaggingMetricsStack } from './TaggingMetricsStack';

export interface Props extends StackProps {
  baseName: string;
  tags: {
    Env: string;
    Team: string;
  };
}


export class PipelineStack extends Stack {
  constructor(app: App, id: string, props: Props) {
    super(app, id, {
      stackName: `${props.baseName}-pipeline-${props.tags.Env}`,
      tags: props.tags,
    });

    const pipeline = new CodePipeline(this, 'pipeline', {
      pipelineName: `${props.baseName}-pipeline-${props.tags.Env}`,
      synth: new ShellStep('Synth', {
        input: CodePipelineSource.gitHub('gplassard/aws-tagging-metrics', 'main', {
          authentication: SecretValue.cfnDynamicReference(new CfnDynamicReference(CfnDynamicReferenceService.SSM, 'github-gplassard-oauth')),
        }),
        commands: ['yarn synth'],
      }),
    });

    const stage = new Stage(this, 'stage', {});
    new TaggingMetricsStack(stage, 'stack', props);
    pipeline.addStage(stage);
  }
}
