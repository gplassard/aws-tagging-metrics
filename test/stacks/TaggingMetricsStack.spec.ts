import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import { TaggingMetricsStack } from '../../src/stacks/TaggingMetricsStack';

describe('TaggingMetricsStack', () => {
  it('should produce the expected cloudformation', () => {
    const app = new App();
    const stack = new TaggingMetricsStack(app, 'TaggingMetricsStack', {
      baseName: 'stack',
      tags: {
        Env: 'dev',
        Team: 'team',
      },
    });

    const template = Template.fromStack(stack);
    expect(template.toJSON()).toMatchSnapshot();
  });
});
