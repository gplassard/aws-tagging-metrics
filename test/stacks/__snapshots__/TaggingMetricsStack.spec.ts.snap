// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaggingMetricsStack should produce the expected cloudformation 1`] = `
{
  "Mappings": {
    "DefaultCrNodeVersionMap": {
      "af-south-1": {
        "value": "nodejs16.x",
      },
      "ap-east-1": {
        "value": "nodejs16.x",
      },
      "ap-northeast-1": {
        "value": "nodejs16.x",
      },
      "ap-northeast-2": {
        "value": "nodejs16.x",
      },
      "ap-northeast-3": {
        "value": "nodejs16.x",
      },
      "ap-south-1": {
        "value": "nodejs16.x",
      },
      "ap-south-2": {
        "value": "nodejs16.x",
      },
      "ap-southeast-1": {
        "value": "nodejs16.x",
      },
      "ap-southeast-2": {
        "value": "nodejs16.x",
      },
      "ap-southeast-3": {
        "value": "nodejs16.x",
      },
      "ca-central-1": {
        "value": "nodejs16.x",
      },
      "cn-north-1": {
        "value": "nodejs16.x",
      },
      "cn-northwest-1": {
        "value": "nodejs16.x",
      },
      "eu-central-1": {
        "value": "nodejs16.x",
      },
      "eu-central-2": {
        "value": "nodejs16.x",
      },
      "eu-north-1": {
        "value": "nodejs16.x",
      },
      "eu-south-1": {
        "value": "nodejs16.x",
      },
      "eu-south-2": {
        "value": "nodejs16.x",
      },
      "eu-west-1": {
        "value": "nodejs16.x",
      },
      "eu-west-2": {
        "value": "nodejs16.x",
      },
      "eu-west-3": {
        "value": "nodejs16.x",
      },
      "me-central-1": {
        "value": "nodejs16.x",
      },
      "me-south-1": {
        "value": "nodejs16.x",
      },
      "sa-east-1": {
        "value": "nodejs16.x",
      },
      "us-east-1": {
        "value": "nodejs16.x",
      },
      "us-east-2": {
        "value": "nodejs16.x",
      },
      "us-gov-east-1": {
        "value": "nodejs16.x",
      },
      "us-gov-west-1": {
        "value": "nodejs16.x",
      },
      "us-iso-east-1": {
        "value": "nodejs14.x",
      },
      "us-iso-west-1": {
        "value": "nodejs14.x",
      },
      "us-isob-east-1": {
        "value": "nodejs14.x",
      },
      "us-west-1": {
        "value": "nodejs16.x",
      },
      "us-west-2": {
        "value": "nodejs16.x",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "summarymetricsexporterE5578E27": {
      "DependsOn": [
        "summarymetricsexporterlogsC0B8B765",
        "summarymetricsexporterServiceRoleDefaultPolicyEA4C9744",
        "summarymetricsexporterServiceRoleFE45D228",
      ],
      "Properties": {
        "Architectures": [
          "arm64",
        ],
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "e776fe06cdbc9b1f402541c322a5321d9b1b49cafdf3a16e3c3a39c7a5f26a23.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "FunctionName": "stack-summary-metrics-exporter-dev",
        "Handler": "index.handler",
        "MemorySize": 512,
        "Role": {
          "Fn::GetAtt": [
            "summarymetricsexporterServiceRoleFE45D228",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "DefaultCrNodeVersionMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 60,
      },
      "Type": "AWS::Lambda::Function",
    },
    "summarymetricsexporterServiceRoleDefaultPolicyEA4C9744": {
      "DependsOn": [
        "summarymetricsexporterlogsC0B8B765",
      ],
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "tag:GetResources",
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "summarymetricsexporterServiceRoleDefaultPolicyEA4C9744",
        "Roles": [
          {
            "Ref": "summarymetricsexporterServiceRoleFE45D228",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "summarymetricsexporterServiceRoleFE45D228": {
      "DependsOn": [
        "summarymetricsexporterlogsC0B8B765",
      ],
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "summarymetricsexporterlogsC0B8B765": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "LogGroupName": "/aws/lambda/stack-summary-metrics-exporter-dev",
        "RetentionInDays": 7,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
